File: java/MongoTest.java
package mongo_api;

import com.mongodb.MongoClient;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Projections;
import com.mongodb.client.model.Sorts;

import java.util.ArrayList;
import java.util.List;

import org.bson.Document;
import org.bson.conversions.Bson;

public class MongoTest {

    public static void main(String[] args) {
        //connect
        MongoClient client = new MongoClient("xxx.xxx.xxx.xx", 27017);

        MongoDatabase db = client.getDatabase("students");

        MongoCollection<Document> collection = db.getCollection("grades");

        Insert(collection);
        Query(collection);
        Update(collection);
        Delete(collection);

        System.out.println("DataBase OK!");
    }

    public static Document CreateDoc() {
        Document doc = new Document().append("student_id", 1200)
                                     .append("type", "hw")
                                     .append("score", 50);
        return doc;
    }

    public static void Insert(MongoCollection<Document> collection) {
        Document doc = CreateDoc();
        collection.insertOne(doc);
        //collection.insertMany(arg0);
    }

    public static void Query(MongoCollection<Document> collection){

        System.out.print("Find One");
        Document doc = collection.find().first();

        System.out.print("Find All into");
        List<Document> doclist = collection.find().into(new ArrayList<Document>());
        for (Document cur: doclist){
            //do some thing
        }

        System.out.print("Find All Iterable");
        MongoCursor<Document> cur = collection.find().iterator();
        try {
            while (cur.hasNext()) {
                Document document = cur.next();
                //do some thing
            }
        } finally {
            cur.close();
        }

        System.out.print("Count");
        long cnt = collection.count();

        //Query with a filter
        Bson filter = Filters.and(Filters.eq("student_id", 5), Filters.gt("score", 45));
        doclist =  collection.find(filter).into(new ArrayList<Document>());
        //Query with projections
        Bson projection = Projections.fields(Projections.include("student_id","score"),
                                                                     Projections.exclude("_id"));
        doclist =  collection.find(filter).projection(projection).into(new ArrayList<Document>());
        //Query with sort,skip,limit
        Bson sort = Sorts.orderBy(Sorts.ascending("student_id","score"), Sorts.descending("type"));
        doclist =  collection.find(filter)
                             .projection(projection)
                             .sort(sort)
                             .limit(10)
                             .skip(5)
                             .into(new ArrayList<Document>());
    }

    public static void Update(MongoCollection<Document> collection){
        Bson filter = Filters.and(Filters.gt("score", 45));
        //replace
        collection.replaceOne(filter, new Document("score",40)
                                                  .append("type", "hw")
                                                  .append("update", true));
        //update
        collection.updateOne(filter, new Document("$set", new Document("score",20)));
        collection.updateMany(filter, new Document("$inc", new Document("score",20)));
    }

    public static void Delete(MongoCollection<Document> collection){
        Bson filter = Filters.and(Filters.gt("score", 45));
        collection.deleteOne(filter);
        collection.deleteMany(filter);
    }
}

File: c_sharp/MongoTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Driver.Core;

namespace mongoexp
{
    class Program
    {
        static void Main(string[] args)
        {
            var connectionString = "mongodb://xxx.xxx.xxx.xx:27017";

            var client = new MongoClient(connectionString);  //connect to mongo server

            var db = client.GetDatabase("students"); //choose database

            var col = db.GetCollection<BsonDocument>("grades"); //choose collection

            BsonDocument doc = testBsonDoc();  //document in C#

            Query(col);
            Console.WriteLine("OK!");
            Console.ReadLine();
        }

        static BsonDocument testBsonDoc()
        {
            var doc = new BsonDocument();
            doc["student_id"] = 10000;
            doc["type"] = "hw2";
            //doc["address"] = new BsonDocument { { "key", "value" } };
            doc.Add("score", 30);

            /*
            var nestedArray = new BsonArray();
            nestedArray.Add("red");
            nestedArray.Add("blue");
            doc["array"] = nestedArray;
            */
            Console.WriteLine(doc);
            return doc;
        }

        static void Insert(IMongoCollection<BsonDocument> col, BsonDocument doc)
        {
            //插入操作，同步异步插入
            col.InsertOne(doc);
            // await col.InsertOneAsync(doc);
            // col.InsertMany(docs);
            // await col.InsertManyAsync(docs);
        }

        static void Query(IMongoCollection<BsonDocument> col)
        {
            //查询，同步，异步
            //Find the First Document in a Collection
            var doc = col.Find(new BsonDocument()).FirstOrDefault();
            //var document = await collection.Find(new BsonDocument()).FirstOrDefaultAsync();
            Console.WriteLine(doc);

            //Find All Documents in a Collection
            var documents = col.Find(new BsonDocument()).ToList();
            foreach(var document in documents)
            {
                Console.WriteLine(document);
            }
            //var documents = await collection.Find(new BsonDocument()).ToListAsync();
            //await collection.Find(new BsonDocument()).ForEachAsync(d => Console.WriteLine(d));

            var cursor = col.Find(new BsonDocument()).ToCursor();
            foreach(var document in cursor.ToEnumerable())
            {
                Console.WriteLine(document);
            }

            //选择条件
            var builder = Builders<BsonDocument>.Filter;
            var filter = builder.Gt("score", 50) & builder.Lte("score", 100);
            cursor = col.Find(filter).ToCursor();

            //排序
            var sortFilter = Builders<BsonDocument>.Sort.Descending("score").Ascending("type");
            cursor = col.Find(filter).Sort(sortFilter).Skip(2).Limit(2).ToCursor();

            //投影
            var projectFilter = Builders<BsonDocument>.Projection.Exclude("_id").Include("type");
            cursor = col.Find(filter).Project(projectFilter).ToCursor();
        }

        static void Update(IMongoCollection<BsonDocument> col)
        {
            var filter = Builders<BsonDocument>.Filter.Lt("score", 100);
            var update = Builders<BsonDocument>.Update.Inc("score", 100);
            var result = col.UpdateOne(filter, update);
            //var result = col.UpdateMany(filter, update);
            if (result.IsModifiedCountAvailable)
            {
                Console.WriteLine(result.ModifiedCount);
            }

            //替换 replacing
            var result_replace = col.ReplaceOne(filter, new BsonDocument("name","yomin"));

        }

        static void Delete(IMongoCollection<BsonDocument> col)
        {
            var filter = Builders<BsonDocument>.Filter.Eq("score", 110));
            var result = col.DeleteOne(filter);
        }
    }
}
